Scala Native 0.4.0

larsrh opened the pull request:
If this PR goes through, I'll back-publish for the last ScalaCheck release.

Should we drop 0.3.9 support then, too?

SethTisue commented:
I vote for dropping 0.3.9. Additional targets are enough of a maintenance burden as-is, without also having to worry about multiple versions of those targets.

larsrh commented:
Turns out they aren't even source compatible:

```
[error] /home/runner/work/scalacheck/scalacheck/native/src/main/scala/org/scalacheck/Platform.scala:14:8: object PreloadedClassLoader is not a member of package scala.scalanative.testinterface
[error] import scala.scalanative.testinterface.PreloadedClassLoader
[error]        ^
[error] /home/runner/work/scalacheck/scalacheck/native/src/main/scala/org/scalacheck/Platform.scala:27:25: not found: type PreloadedClassLoader
[error]     loader.asInstanceOf[PreloadedClassLoader].loadPreloaded(name)
[error]                         ^
[error] /home/runner/work/scalacheck/scalacheck/native/src/main/scala/org/scalacheck/Platform.scala:14:40: Unused import
[error] import scala.scalanative.testinterface.PreloadedClassLoader
[error]                                        ^
```

So either I'll have to cheat when back-publishing, or cry.

gabro commented:
Thanks @larsrh for working on this, this should also unblock support for 0.4.0 support in MUnit (across 2.11, 2.12 and 2.13)!

About dropping 0.3.9, I think it would be great to nudge people out of the 2.11 ecosystem, so that we can finally move on (like in Metals, for example, where 2.11 support is largely there for Scala Native projects)

larsrh commented:
@WojciechMazur if you could give us some pointers how this code should look like nowadays, it would be much appreciated!

gabro commented:
@larsrh I'm looking at similar errors in MUnit and probably this helps https://github.com/scala-native/scala-native/commit/3077e74c363307963ce6dfbe7da1969bb76993b3

gabro commented:
For reference, here's the equivalent fix in MUnit (just tested, it runs correctly) https://github.com/scalameta/munit/pull/298/files#diff-16de51ca52ec70a26d1950ea7aa6c11432074b200e28d9316cba1077dbaadf3a

WojciechMazur commented:
I also personally would suggest dropping 0.3.x. Instead of using `PreloadedClassLoader` directly in 0.3.9 you could use `org.scalajs.testinterface.TestUtils` (deprecated since 0.4.0), it's boundled in test-interface

larsrh commented:
I've dropped Scala Native 0.3.9, Scala 2.11, changed the build accordingly, but left the broken code intact. To be honest, I don't understand what I need to change in there. If anyone could give me a patch, that'd be much appreciated.

gabro commented:
I can take a look.

About 2.11, why the drop?

larsrh commented:
Why not? Scala Native is the only reason why anyone still supports it, and that reason just got void.

gabro commented:
I agree in principle, but I wouldn’t conflate the two.
It would mean to drop 2.11 downstream as well (e.g MUnit-scalacheck) and I
would prefer to do it separately than this update

On Thu, 21 Jan 2021 at 20:44 Lars Hupel <notifications@github.com> wrote:

> Why not? Scala Native is the only reason why anyone still supports it, and
> that reason just got void.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/typelevel/scalacheck/pull/751#issuecomment-764892335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFI5ZGGWCMRERD5O2M7QWTS3B73FANCNFSM4WJQEOQA>
> .
>
-- 

<https://buildo.io/>

gabriele petronella / co-founder
gabriele@buildo. <gabriele@buildo.io>

buildo
https://buildo.io

larsrh commented:
Sounds plausible, but I was hoping to sneakily back-publish Scala Native support for the latest ScalaCheck release, so you won't even notice downstream :wink:

SethTisue commented:
> About 2.11, why the drop?

my take on this is: because supporting old versions of things is death by a thousand papercuts for unpaid OSS maintainers

gabro commented:
@SethTisue 100% agree, I was just wondering on the opportunity of doing so while also dropping a target version at the same time.

But back-publishing resolves this, so no objections on my end 🚀

@larsrh in the meantime, I made it compile, but I'm hitting a runtime crash when running the tests due to how the new reflective instantiation works.

I'll dig a bit more into it later today

gabro commented:
All green ✅ 🎉

larsrh commented:
Thanks a ton @gabro!

larsrh commented:
Seems like it also builds for 1.15.2. Running the release right now.

larsrh commented:
Finally the artifacts appear to have been released. Sonatype was exceptionally slow today.

gabro commented:
Yep, I can confirm I've successfully downloaded them in MUnit, thanks!
